### PR TITLE
fix(mcp): read the auth context as the CLJS map it is

### PR DESCRIPTION
**Pull request by riatzukiza**
## What is broken

The consent page 500s on the **first authenticated request** of the OAuth flow,
so no client can get past login:

```
500 {"statusCode":500,"error":"Internal Server Error",
     "message":"Cannot read properties of undefined (reading 'email')"}
```

Production stack trace:

```
TypeError: Cannot read properties of undefined (reading 'email')
    at knoxx.backend.infra.routes.mcp.js:732:123
    at authorization_consent_html (knoxx.backend.infra.routes.mcp.js:743:3)
    at Object.handler (knoxx.backend.infra.routes.mcp.js:1109:43)
```

## Why

`resolve-auth-context` returns a **ClojureScript map** — `policy-db/resolve-context!`
builds it with keyword keys, which is why every other route reads it through
`authz/ctx-user-email` and friends. This module alone reached in with `aget`:

```clojure
(or (aget auth-context "user" "email") (aget auth-context "userEmail") "")
```

That compiles to `ctx["user"]["email"]`. On a CLJS map the intermediate is
`undefined`, so the nested read throws *before* the `or` can fall through — the
fallback that was meant to make this shape-tolerant never gets to run.

## The change

Five sites move to the shared accessors:

| site | read |
|---|---|
| consent page | email, org slug |
| `authorize/confirm` | membership id, email, org slug |
| `GET /api/mcp/tokens` | membership id |
| `DELETE /api/mcp/tokens/:tokenId` | membership id |

The last two carried the identical expression and would have failed the same way
as soon as anyone reached them.

## Why nothing caught it

This is the first step that requires a real session. The browser-auth guard
`302`s an unauthenticated caller to login *before* this code runs, so the
unauthenticated probes used to verify #212 and #213 cannot reach it by
construction. Three bugs in this flow have now hidden behind that guard.

The new test drives the real authorize route with a real-shaped CLJS context and
asserts the page renders with the user's email and org in it. Reverting the fix
fails it with the production error:

```
FAIL in (consent-page-renders-from-a-cljs-auth-context)
expected: (= :ok outcome)
  actual: (not (= :ok #object[TypeError TypeError: Cannot read properties of undefined (reading 'email')]))
```

It catches the rejection rather than awaiting it bare — the regression throws,
and an unhandled rejection takes the whole runner down mid-suite instead of
reporting one failed test.

691 tests / 2013 assertions, 0 failures. `clj-kondo` clean.

## Still unverified after this

The consent page is where the flow now resumes. Beyond it sit the confirm
redirect, the PKCE code exchange and the first authenticated `POST /mcp` — none
of which have ever been exercised end to end, for the same guard reason. Expect
that a real connector attempt is the only thing that can clear them.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved authorization handling for MCP consent pages, authorization codes, and token management.
  * Ensured user and organization details are displayed correctly during authenticated authorization flows.

* **Tests**
  * Added regression coverage for authenticated consent-page rendering, including email and organization information.

### CODE REVIEW

File: backend/src/cljs/knoxx/backend/infra/routes/mcp.cljs [hunk 1]
```diff
@@ -505,9 +511,10 @@
       (let [requested (requested-tools runtime config auth-context selected-tools)]
         (when (empty? requested)
           (throw (http-error 400 "invalid_scope" "No valid tools selected")))
-        (let [membership-id (str (or (aget auth-context "membership" "id") (aget auth-context "membershipId") ""))
-              user-email    (str (or (aget auth-context "user" "email") (aget auth-context "userEmail") ""))
-              org-slug      (str (or (aget auth-context "org" "slug") (aget auth-context "orgSlug") ""))
+        ;; Same CLJS-map accessors as the consent page; see the note there.
+        (let [membership-id (str (or (authz/ctx-membership-id auth-context) ""))
+              user-email    (str (or (authz/ctx-user-email auth-context) ""))
+              org-slug      (str (or (authz/ctx-org-slug auth-context) ""))
```
[the remark this answers is not in the document]
Comment (riatzukiza) [reply backend/src/cljs/knoxx/backend/infra/routes/mcp.cljs:517]: Fixed in 84eb2a50. Valid — and the consequence reaches further than the code document: `persist-access-token!` copies `membershipId`/`userEmail` verbatim out of the code record into the access token, so a blank membership became a valid bearer token carrying no authorization identity.

Membership and user email are now required before the payload is built, returning `400 missing_identity`. The empty-string fallbacks stay for the render path, where a partly-resolved context must not crash.

I deliberately did **not** require `org-slug`. It is descriptive rather than an authorization key, and a membership without one is legitimate — making it mandatory risks 400ing real logins in a flow that has already broken three times. Say if you want it required too.

File: backend/src/cljs/knoxx/backend/infra/routes/mcp.cljs [hunk 2]
```diff
@@ -567,7 +574,7 @@
 (defroute mcp-revoke-user-token! [browser-auth-guard] "DELETE" "/api/mcp/tokens/:tokenId" [browser-auth-guard]
   (let [auth-context  (aget request "authContext")
         {:keys [token-id]}    (parse-revoke-token-params request)
-        membership-id         (str (or (aget auth-context "membership" "id") (aget auth-context "membershipId") ""))]
+        membership-id         (str (or (authz/ctx-membership-id auth-context) ""))]
     (when (or (str/blank? membership-id) (str/blank? token-id))
       (throw (http-error 400 "invalid_request" "membership and tokenId are required")))
     (await (mongo-mcp/delete-token! token-id))
```
[the remark this answers is not in the document]
Comment (riatzukiza) [reply backend/src/cljs/knoxx/backend/infra/routes/mcp.cljs]: Fixed in 84eb2a50. This was the more serious of the two, and it was exploitable: the listing route is per-membership, but nothing stopped a hand-made `DELETE /api/mcp/tokens/:tokenId` with someone else's token value.

Added `delete-token-for-membership!`, which matches on `access_token` **and** `membership_id` and returns whether anything was deleted. The route now answers `404` on a miss instead of a false `200`. Since the query is membership-scoped, that 404 reveals nothing about whether the token exists for another membership.

`delete-token!` stays for callers that legitimately act outside a membership, with its lack of scoping now documented rather than incidental.

Tested at the store level, since that is where the predicate lives:

| case | result |
|---|---|
| revoke own token | deleted, returns true |
| revoke another membership's | **not** deleted, returns false |
| revoke unknown id | returns false |

Removing the membership predicate fails two of those, with the other membership's token actually gone — so the test reproduces the vulnerability rather than just asserting the fix.

695 tests / 2022 assertions, 0 failures.

File: backend/src/cljs/knoxx/backend/infra/stores/mongo_mcp_oauth.cljs [hunk 3]
```diff
@@ -164,14 +164,34 @@
...
+   caller can tell 'revoked' from 'not yours, or not there'. Deleting by token
+   value alone lets any authenticated caller revoke another membership's token
+   if they learn its value."
+  ([access-token membership-id]
+   (delete-token-for-membership! (mongo-client/get-db) access-token membership-id))
+  ([db access-token membership-id]
+   (when (and db access-token membership-id)
+     (let [c (tokens-coll db)
+           result (await (.deleteOne c #js {"access_token"  (str access-token)
+                                            "membership_id" (str membership-id)}))]
```
[the remark this answers is not in the document]
Comment (riatzukiza) [reply backend/src/cljs/knoxx/backend/infra/stores/mongo_mcp_oauth.cljs]: Fixed in 92694092 — grounded, and cheaper to honour than it looked. `knoxx.backend.extern.mongo` already owns this exact boundary ("callers inject a native collection handle... all raw interop with that handle is born and dies here"); it simply had no delete, so my `#js` query and `aget` on `DeleteResult` had pushed the boundary up into `infra.stores`.

Added `extern-mongo/delete-one!`, taking a CLJS query and returning `{:deleted-count n}`. The driver's native result no longer escapes, and the revocation path now contains **zero** raw interop expressions.

Per the AGENTS.md instruction to add a conversion regression test when an extern adapter grows a boundary, there are two: `DeleteResult` decodes to CLJS with the query correctly encoded, and a result missing `deletedCount` decodes to `0` rather than `nil`.

Scoped deliberately to what this branch adds. The other ten functions in this store do raw interop directly and predate the branch — converting them is a worthwhile follow-up, but not something I'll attach to a security fix in a flow that has already broken three times in production.

File: backend/src/cljs/knoxx/backend/infra/stores/mongo_mcp_oauth.cljs [hunk 4]
```diff
@@ -164,14 +164,34 @@
...
+  "Delete an access token only when it belongs to this membership.
+
+   Returns true when a token was deleted and false when nothing matched, so a
+   caller can tell 'revoked' from 'not yours, or not there'. Deleting by token
+   value alone lets any authenticated caller revoke another membership's token
+   if they learn its value."
+  ([access-token membership-id]
+   (delete-token-for-membership! (mongo-client/get-db) access-token membership-id))
+  ([db access-token membership-id]
+   (when (and db access-token membership-id)
```
[the remark this answers is not in the document]
Comment (riatzukiza) [reply backend/src/cljs/knoxx/backend/infra/stores/mongo_mcp_oauth.cljs]: Fixed in 92694092. Contracts are now named at the boundary rather than left to the calling route:

- `RevocationRequest` — non-blank `:access-token` and `:membership-id`
- `RevocationResult` — `:deleted-count` a non-negative int

Blank identity returns `false` **without issuing a delete**, instead of falling through to a query with an empty-string field. That matches nothing against today's schema, but it is one schema change away from matching more than intended, and as you say the obligation belongs here rather than in whichever route happens to call in. A result that fails `RevocationResult` throws rather than quietly reading as "nothing deleted".

The blank-identity test asserts against a collection double that records whether `deleteOne` was reached at all, so it proves the delete is never issued rather than merely that it found nothing.

Worth flagging honestly: no other store in this codebase names a Malli validator, so this is the first. I took that as the direction AGENTS.md intends for new code rather than a reason to skip it — but if you'd rather stores stay uncontracted until there's a shared pattern, say so and I'll reconsider.

698 tests / 2030 assertions, 0 failures; JS boundary gate passes.

File: backend/src/cljs/knoxx/backend/infra/stores/mongo_mcp_oauth.cljs [hunk 5]
```diff
@@ -164,14 +166,56 @@
...
    (when (and db access-token)
      (let [c (tokens-coll db)]
        (await (.deleteOne c #js {"access_token" (str access-token)}))
        true))))
 
+;; Contracts for the token-revocation boundary. Both obligations are named
+;; here rather than left to whichever route happens to call in: an identity
+;; that is present but blank must not be able to widen the delete, and a
+;; decoded result that is missing its count must not read as "nothing deleted".
+(def RevocationRequest
```
[the remark this answers is not in the document]
Comment (riatzukiza) [reply backend/src/cljs/knoxx/backend/infra/stores/mongo_mcp_oauth.cljs]: Fixed in 2e911f6e. Correct — the category split puts contracts in `law.*` with no I/O, and I had them sitting in the effectful store.

Moved to `knoxx.backend.law.mcp-oauth`. `law.openplanner-translation` turned out to be the exact precedent — Malli contracts for a Knoxx Mongo boundary, already in `law.*` — so this mirrors it rather than inventing a shape. Worth noting `law.store` was *not* the right home despite the name: it is a re-export shim over the contract-runtime store law module.

The store now names the obligations without owning them; zero schema definitions remain in it. Added direct coverage of the contracts while they had a home worth testing — including that whitespace is not an identity, and that a result which lost its count raises rather than reading as "nothing deleted". That throw path was previously unexercised.

700 tests / 2038 assertions, 0 failures; JS boundary gate passes; clj-kondo clean.

Thanks for the three-step pull on this one — the sequence (scope the delete, get the interop behind the adapter, then get the contract out of the store) landed somewhere better than any single step would have.

File: backend/src/cljs/knoxx/backend/extern/mongo.cljs [hunk 6]
```diff
@@ -12,6 +12,17 @@
...
+(defn ^:async delete-one!
+  "Delete at most one document matching a CLJS field-equality query.
+
+   Returns {:deleted-count n} as CLJS data. The driver's native DeleteResult is
+   decoded here and never escapes, so callers can act on the count without
+   knowing the SDK shape — a caller that reads .deletedCount itself has moved
+   the boundary upstream."
+  [collection-handle query]
+  (let [result (await (.deleteOne collection-handle (clj->js query)))]
+    {:deleted-count (or (aget result "deletedCount") 0)}))
```
[the remark this answers is not in the document]
Comment (riatzukiza) [reply backend/src/cljs/knoxx/backend/extern/mongo.cljs]: Fixed in 6f8d570b. You caught a contradiction I had written into my own code: `law.mcp-oauth` states that a result which lost its count must not be readable as "nothing was deleted", and then `delete-one!` substituted zero for a missing `deletedCount` — so the contract could never see one. The fabrication validated, and the route answered a confident 404 for a persistence layer it had not actually managed to read.

`delete-one!` now decodes faithfully:

| driver result | decoded |
|---|---|
| `{deletedCount: 1}` | `{:deleted-count 1}` |
| `{deletedCount: 0}` | `{:deleted-count 0}` — genuine miss, still a count |
| `{acknowledged: true}` | `{:deleted-count nil}` → contract fails → store raises |
| `{deletedCount: "1"}` | `{:deleted-count nil}` — not coerced |

Keeping a genuine `0` distinct from "no count reported" is the point: "nothing matched" and "I could not read the answer" must not collapse into the same 404.

The previous revision had a test asserting the fabrication was *correct*. That test is now inverted, and the throw case runs end to end through the real adapter rather than stubbing it, so the adapter and the contract cannot drift apart again. Reverting the decode fails three assertions including the throw case.

700 tests / 2040 assertions, 0 failures. The revocation path is `delete-one!`'s only caller, so nothing else depended on the old zero.
